File: 2045-second-minimum-time-to-reach-destination/2045-second-minimum-time-to-reach-destination.py
class Solution:
    def secondMinimum(self, n: int, edges: List[List[int]], tt: int, change: int) -> int:
        g = defaultdict(set)

        for u, v in edges:
            g[u].add(v)
            g[v].add(u)
        
        q = {1}
        at = 0
        dt = 0
        vis_cnt = Counter()
        
        while q:
            sz = len(q)
            nxt = set()
            for _ in range(sz):
                node = q.pop()
                if node == n and vis_cnt[node] == 1:
                    return at
                if vis_cnt[node] == 2:
                    continue
                vis_cnt[node] += 1
                for nei in g[node]:
                    if vis_cnt[nei] < 2:
                        nxt.add(nei)

            q = nxt
            at = dt + tt
            cycle, _ = divmod(at, change)
            
            if cycle % 2 == 1:
                dt = change * (cycle + 1)
            else:
                dt = at
        
        return -1

File: 2045-second-minimum-time-to-reach-destination/2045-second-minimum-time-to-reach-destination.cpp
class Solution {
public:
    int secondMinimum(int n, vector<vector<int>>& edges, int tt, int change) {
        unordered_map <int, unordered_set <int>> g;

        for (auto edge : edges) {
            int u = edge[0], v = edge[1];
            g[u].insert(v);
            g[v].insert(u);
        }
        
        queue <int> q;
        q.push(1);
        int at = 0;
        int dt = 0;
        unordered_map <int, int> vis_cnt;

        while (!q.empty()) {
            int sz = q.size();
            unordered_set <int> nxt;
            for (int i = 0; i < sz; i++) {
                int node = q.front(); q.pop();
                if (node == n && vis_cnt[node] == 1) {
                    return at;
                }
                if (vis_cnt[node] == 2) {
                    continue;
                }
                vis_cnt[node]++;
                for (int nei : g[node]) {
                    if (vis_cnt[nei] < 2) {
                        nxt.insert(nei);
                    }
                }
            }
            for (int next_node : nxt) {
                q.push(next_node);
            }
            at = dt + tt;
            int cycle = at / change;
            
            if (cycle % 2 == 1) {
                dt = change * (cycle + 1);
            } else {
                dt = at;
            }
        }
        
        return -1;
    }
};